### PR TITLE
Blockchain sync resilence (SEC-021, SEC-026, SEC-023)

### LINKED ISSUE #2896: SEC-026 Unbounded Memory Usage in Block Queue

### PR DESCRIPTION
**Pull request by arkpar**
Fixes #2896 #2893 #2891 (partially)

### CODE REVIEW

File: libethereum/BlockChainSync.cpp [hunk 1]
```diff
@@ -298,6 +299,12 @@ void BlockChainSync::onPeerNewBlock(std::shared_ptr<EthereumPeer> _peer, RLP con
 		case ImportResult::FutureTimeUnknown:
 		case ImportResult::UnknownParent:
 		{
+			if (++_peer->m_unknownNewBlocks > c_maxPeerUknownNewBlocks)
```
Comment (gavofyork) [inline libethereum/BlockChainSync.cpp]: anyone else to who can't immediately tell whether `++foo->` binds first left or right? or is it just me?

File: libethereum/EthereumPeer.cpp [hunk 2]
```diff
@@ -304,6 +307,11 @@ bool EthereumPeer::interpret(unsigned _id, RLP const& _r)
 			break;
 		}
 		setIdle();
+		if (itemCount > c_maxIncomingHashes)
+		{
+			disable("Too many hashes");
```
Comment (gavofyork) [inline libethereum/EthereumPeer.cpp]: future PR might check simply whether the `itemCount` is at most the number we actually asked for.

File: libethereum/EthereumPeer.cpp [hunk 3]
```diff
@@ -364,8 +372,15 @@ bool EthereumPeer::interpret(unsigned _id, RLP const& _r)
 	case NewBlockHashesPacket:
 	{
 		unsigned itemCount = _r.itemCount();
+
 		clog(NetMessageSummary) << "BlockHashes (" << dec << itemCount << "entries)" << (itemCount ? "" : ": NoMoreHashes");
 
+		if (itemCount > c_maxIncomingNewHashes)
```
Comment (gavofyork) [inline libethereum/EthereumPeer.cpp]: there's not technically a limit on this in the protocol. we should place a limit (e.g. 32) and make other clients aware of it.

File: libethereum/TransactionQueue.cpp [hunk 4]
```diff
@@ -31,6 +31,8 @@ using namespace dev::eth;
 const char* TransactionQueueChannel::name() { return EthCyan "┉┅▶"; }
 const char* TransactionQueueTraceChannel::name() { return EthCyan " ┅▶"; }
 
+const size_t c_maxVeririficationQueueSize = 8192;
```
Comment (gavofyork) [inline libethereum/TransactionQueue.cpp]: one too many syllables in "verirification" :)

File: libethereum/TransactionQueue.cpp [hunk 5]
```diff
@@ -370,13 +372,23 @@ void TransactionQueue::clear()
...
 {
+	bool queued = false;
 	{
 		Guard l(x_queue);
 		unsigned itemCount = _data.itemCount();
 		for (unsigned i = 0; i < itemCount; ++i)
+		{
+			if (m_unverified.size() >= c_maxVeririficationQueueSize)
+			{
+				clog(TransactionQueueChannel) << "Transaction verification queue is full. Dropping" << itemCount - i << "transactions";
```
Comment (gavofyork) [inline libethereum/TransactionQueue.cpp]: this just pops later transactions, right? so still attackable by a constant stream of useless transactions? might be worth sometimes culling from the queue itself.

Comment (arkpar) [inline libethereum/TransactionQueue.cpp]: For malfromed transactions the peer who sends them will eventually loose rating (and presumably banned) here: https://github.com/ethereum/cpp-ethereum/blob/develop/libethereum/EthereumHost.cpp#L386
For well-formed transactions It is more complicated. For instance an attacker could send a stream of perfectly fine transactions, with high gas price from an account with zero balance. Such transactions will only fail at execution phase.  Would it be correct to lower rating for peers that propagate them? Currently peers propagate transactions without checking if they execute first.

File: libethereum/EthereumPeer.cpp (hunk 2 above)
Comment (arkpar) [inline libethereum/EthereumPeer.cpp]: A quick change, pushed.

**Pull request merged**